---
title: "GeneSLand"
author: "Erik Bot"
date: "2024-01-24"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{GeneSLand}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette is a quick guide to use the `GeneSLand` package. 
GeneSLand (Gene Specificity Landscape) is an R package for comparative transcriptomic analysis following the framework described in Bot and Davila-Velderrain, 2025.

The package can be installed by:

```{r, eval=FALSE}
install.packages("devtools")

devtools::install_github("https://github.com/davilavelderrainlab/GeneSLand")
```

And then loaded:

```{r setup}
library(GeneSLand)
```

```{r style, echo = FALSE, results = 'asis'}
library(BiocStyle)
```

```{r, echo = FALSE}
library(knitr)
```

To create a Specificity Single Cell Experiment object (`SSCE`), we can provide either a bulk RNAseq dataset, or first create a pseudobulk from a scRNAseq dataset by `get_pseudobulk`.

For this vignette, we will use a randomly generated dataset, which has to be substituted by your own dataset.

```{r}
set.seed(123)
n_profiles <- 50
n_genes <- 5000
expression_dataset <- matrix(runif(n_genes*n_profiles,1,10), ncol = n_profiles)
colnames(expression_dataset) <- paste0('Cell-', seq(1,dim(expression_dataset)[2]))
rownames(expression_dataset) <- paste0('Gene-', seq(1,dim(expression_dataset)[1]))
```

To create the SSCE, just run `createSSCE`:
 
```{r, results='hide', message=FALSE}
ssce <- createSSCE(xProfiles = expression_dataset)
```

The obtained SSCE now contains several informations: under `assays` there are, for each gene, the expression breadth (B) score for the 50 expression level (L) bins. Additionally, the `AUC`, `dRate`, `lbSpec` and initial B scores can be easily accessed by:

```{r, eval=FALSE}
#AUC
ssce$AUC

#dRate
ssce$dRate

#lbSpec
ssce$lbSpec

#initial B score
ssce$initB
```

By using `createSSCE` we also built the 2D gene specificity landscape, which we can easily visualize by `plotSSCE`, along with the specificity scores:

```{r}
plotSSCE(ssce)
# AUC
plotSSCE(ssce, values = 'AUC')
# dRate
plotSSCE(ssce, values = 'dRate')
# lbSpec
plotSSCE(ssce, values = 'lbSpec')
```

In the plot each dot is a gene, and the vicinity in the plot indicates a similar expression behavior. 
When the specificity scores are shown, a darker color is given to genes with lower values, whereas a brighter color corresponds to genes with high values.

We can also plot the LB lines for a gene or multiple genes:

```{r}
plot_lines_SSCE(ssce, gene = 'Gene-1')
plot_multiple_lines_SSCE(ssce, genes = c('Gene-1', 'Gene-2'), genes_colors = c('red', 'blue'))
```

The line shows how many cell types/tissues express the gene more than the expression level bin. 
So, if for Gene-1 the line has a B value of 0.5 at L = 20, it means that 50% of the cell types express Gene-1 more than 20. 
The bin values (like 20 in this case) are not counts, but obtained by dividing the expression range in the dataset (from maximum to minimum expression) in 50 equally distant parts. 20 indicates the expression value that is 20/50 of the maximum expression value.

# Gene groups SSCE

We can also create SSCEs starting from gene sets of choice, like Gene Ontology Biological Processes, or HUGO gene groups. To do so, we just need to provide a list of gene sets, in which each element is a gene set and contains a series of genes.

For this example, we just randomly divided our genes into 26 groups.

```{r}
gene_groups <- split(rownames(expression_dataset), sample(LETTERS, nrow(expression_dataset), replace = TRUE))
ssce_groups <- createSSCE(expression_dataset, geneGroups = gene_groups, nRand = 20)
```

If you want to add the scores significance or the intra-group variability after creating the SSCE, 
you can by using `compute_significance_SSCE` and `compute_variability_SSCE`:

```{r}
ssce_groups <- compute_significance_SSCE(ssce_groups)
ssce_groups <- compute_variability_SSCE(ssce_groups)
```

The score significance indicates how statistically significant are the scores compared to a random gene sets with the same number of genes, whereas the variability shows how homogeneous the gene set is in terms of individual scores.

As we did for single genes, we can now plot the related LB lines:

```{r}
plot_lines_SSCE(ssce_groups, gene = 'Z')
plot_multiple_lines_SSCE(ssce_groups, genes = c('Z', 'A'), genes_colors = c('red', 'blue'))
```

Now the plot shows also the intra-gene set variability (dashed black line), but also the random distribution in grey, along with its variability (dashed grey line).

# Map new lines on existing SSCEs

If we want to compare the expression behavior of genes in different contexts, an easy visualization can be obtained by mapping the LB behavior of genes onto an existing SSCE and see how they deviate. To do so, we just need to run the `mapSSCE` function giving a SSCE, the new expression profiles and the wanted gene or genes. Also in this case, we will generate the new profiles randomly.

```{r}
set.seed(123)
n_profiles <- 50
n_genes <- 5000
new_profiles <- matrix(runif(n_genes*n_profiles,1,10), ncol = n_profiles)
colnames(new_profiles) <- paste0('NewCell-', seq(1,dim(new_profiles)[2]))
rownames(new_profiles) <- paste0('Gene-', seq(1,dim(new_profiles)[1]))
```

We can now map the new profiles:

```{r}
mapped_ssce <- mapSSCE(new_profiles = new_profiles, ssce = ssce, genes = 'Gene-1')
```

In the `mapped_ssce` object we now have the new coordinates of the gene (under `Coordinates`), the `LineObject`, with the B values and the AUC, dRate and lbSpec scores, and the plot, which can be visualized by:

```{r}
mapped_ssce$Plot
```

From the plot we can easily see where the new gene (or genes) end up compared to the previous context.

```{r}
sessionInfo()
```

